### PR TITLE
Added JSON lightcurve generation functionality.

### PR DESCRIPTION
**Pull request by hughdickinson**
Hi Adam.

Absolutely **no rush** on this, but would appreciate any feedback when you have time. Just my first attempt at implementing some JSON generation. Will be useful for the Black Hole Hunters project and maybe also for you when you make your dynamic folding viewer.

Cheers,

Hugh

### CODE REVIEW

File: starcatalogue/models.py [hunk 1]
```diff
@@ -57,8 +60,30 @@ def get_or_generate_image(self, image_attr, generation_task, default=None):
...
+            if (
+                not self.json_files_celery_task_id
+                or AsyncResult(self.json_files_celery_task_id).ready()
+                or not self.json_file_celery_started
+                or self.json_file_celery_started < (timezone.now() - datetime.timedelta(minutes=5))
+            ):
+                self.json_files_celery_task_id = generation_task.delay(self.id).id
+                self.json_file_celery_started = timezone.now()
+                self.save()
+            if image_not_present:
```
Comment (adammcmaster) [inline starcatalogue/models.py:77]:
```suggestion
            if json_not_present:
```

File: starcatalogue/models.py [hunk 2]
```diff
@@ -72,6 +97,11 @@ class Star(models.Model, ImageGenerator):
     image_version = models.FloatField(null=True)
     image_celery_started = models.DateTimeField(null=True)
 
+    json_file = models.ImageField(null=True, upload_to=star_json_upload_to)
```
Comment (adammcmaster) [inline starcatalogue/models.py:100]:
```suggestion
    json_file = models.FileField(null=True, upload_to=star_json_upload_to)
```

File: starcatalogue/models.py [hunk 3]
```diff
@@ -184,22 +214,29 @@ def image_location(self):
     def get_image_location(self):
         return self.get_or_generate_image(self.image_file, generate_star_images)
 
+    @property
+    def json_location(self):
+        return self.get_json_location()
+
+    def get_json_location(self):
+        return self.get_or_generate_json(self.image_file, generate_star_json_files)
```
Comment (adammcmaster) [inline starcatalogue/models.py:222]:
```suggestion
        return self.get_or_generate_json(self.json_file, generate_star_json_files)
```

File: starcatalogue/tasks.py [hunk 4]
```diff
@@ -3,6 +3,7 @@
 import urllib
 import yaml
 import zipfile
+import json
```
Comment (adammcmaster) [inline starcatalogue/tasks.py:6]:
```suggestion
import ujson as json
```

ujson is faster (we'll also need to install it with pipenv).

File: starcatalogue/tasks.py [hunk 5]
```diff
@@ -77,13 +78,40 @@ def download_fits(star_id):
...
+
+    if not star.fits:
+        return
+
+    ts = star.timeseries
+    if not ts:
+        return
+    ts_flux = Star.outlier_clip(ts["TAMFLUX2"])
+    ts_data = {
+        "data": {"x": ts.time.jd, "y": ts_flux},
```
Comment (adammcmaster) [inline starcatalogue/tasks.py:93]: I'm not sure if these are actually directly JSON serialisable, but I guess let's try it and see!

File: starcatalogue/tasks.py [hunk 6]
```diff
@@ -77,13 +78,40 @@ def download_fits(star_id):
...
+        "chartOptions": {
```
Comment (adammcmaster) [inline starcatalogue/tasks.py:94]: Are these options used by the Zooniverse front end? Maybe add a comment with a link to docs if there are any.

**Review by adammcmaster — CHANGES_REQUESTED** [review]
This is a good start. I've suggested some minor changes.

We'll also need to generate migrations and install ujson:

```
docker-compose build && docker-compose up -d
docker-compose exec app python manage.py makemigrations
docker-compose exec app pipenv install ujson
docker-compose stop
```

File: starcatalogue/tasks.py (hunk 6 above)
Comment (hughdickinson) [reply starcatalogue/tasks.py:94]: Good idea - will do.

File: starcatalogue/tasks.py (hunk 5 above)
Comment (hughdickinson) [reply starcatalogue/tasks.py:95]: If not, can just convert to lists.

**Comment by adammcmaster** [comment]
fyi you might want to rebase this to pull in ede9aaefb1bba48cb87ba74c344b981638a1a73e -- will make local testing easier with the Celery tasks running more often.

**Review by adammcmaster — APPROVED** [review]

**Pull request merged**